File: acmicpc/acmicpc10026.py
import sys
from collections import deque

sys.setrecursionlimit(100000)

drows = [-1, 1, 0, 0]
dcols = [0, 0, -1, 1]

def dfs(sPoint, color):
    global visited, n, graph
    if visited[sPoint[0]][sPoint[1]] or graph[sPoint[0]][sPoint[1]] != color:
        return
    else:
        visited[sPoint[0]][sPoint[1]] = 1

    for i in range(4):
        dr = sPoint[0] + drows[i]
        dc = sPoint[1] + dcols[i]

        if (0 <= dr < n) and (0 <= dc < n):
            dfs([dr, dc], color)

    return


if __name__ == "__main__":
    global visited, n, human

    # sys.stdin = open('sample_input.txt', 'r')

    human = 0
    cow = 0
    
    n = int(sys.stdin.readline())
    graph = []
    for _ in range(n):
        graph.append(list(sys.stdin.readline().strip()))

    visited = [[0] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                dfs([i, j], graph[i][j])
                human += 1

    for i in range(n):
        for j in range(n):
            if graph[i][j] == 'R':
                graph[i][j] = 'G'

    visited = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                dfs([i, j], graph[i][j])
                cow += 1

    print(human, cow)

File: acmicpc/acmicpc10026.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int n;
vector<string> graph;
vector< vector<int> > visited;
vector<int> temp;

vector<int> drows = { -1, 1, 0, 0 };
vector<int> dcols = { 0, 0, -1, 1 };

int dr, dc;
char color_graph;
void dfs(int r, int c, char color) {
    color_graph = graph[r][c];
    if (visited[r][c] || color_graph != color) {
        return;
    }
    else {
        visited[r][c] = 1;
    }

    for (int i=0; i < n; i++) {
        dr = r + drows[i];
        dc = c + dcols[i];

        if (0 <= dr && dr < n && 0 <= dc && dc < n) { 
            if (!visited[dr][dc]) {
                dfs(dr, dc, color);
            }
        }
    }

}

int main() {
    // freopen("sample_input.txt", "r", stdin);

    int human, cow;
    human = 0;
    cow = 0;

    cin >> n;

    string s;
    for (int i=0; i < n; i++) {
        cin >> s;
        graph.push_back(s);
        temp.clear();
        for (int j=0; j < n; j++) {
            temp.push_back(0);
        }
        visited.push_back(temp);
    }

    for (int i=0; i < n; i++) {
        for (int j=0; j < n; j++) {
            if (!visited[i][j]) {
                dfs(i, j, graph[i][j]);
                human++;
            }
        }
    }
    char c;
    for (int i=0; i < n; i++) {
        for (int j=0; j < n; j++) {
            visited[i][j] = 0;
            c = graph[i][j];
            if (c == 'R') { graph[i][j] = 'G'; }
        }
    }

    for (int i=0; i < n; i++) {
        for (int j=0; j < n; j++) {
            if (!visited[i][j]) {
                dfs(i, j, graph[i][j]);
                cow++;
            }
        }
    }

    cout << human << " " << cow << "\n";
}